### ISSUE TITLE
ASAP!! [Waves Exchange]- [condition orders] - Waiting almost 2 month for a solution to our problem :(

### ISSUE THREAD

**Issue by NikBuka** [ccxt/ccxt#16996]
### ؜

Hi, this is the Waves Exchange team. 
Since February 6, 2023 I have tried reaching your team (specifically Igor Kroitor, Carlos, Carlo) in our general telegram chat https://t.me/c/1670109218/94 and by email. Unfortunately, it appears things have gone quiet on your end. 

It has already been almost a month since my first inquiry, so I am writing you a ticket here and hope to receive an answer as soon as possible as we are planning a release this week.

Here’s the main point of my inquiry: 
Our team has developed limit orders and certain updates are required in the CCXT library to support this feature. 
Orders now have the 'attachment' field (which will be included in signatures in the future, but not yet). The format is base58. The charset is UTF-8. In its current state, we only transfer the conditions for stop-limit / stop-market orders. The structure format is as below (use json no-spaces): 

-----------------------------------------------------
{
  "v": 1, //version (int)
  "c": { //condition (object)
    "t": "sp", //condition type. for now only "stop-price" (string)
    "v": { //value (object)
      "p": 123123 //price (long)
    }
  }
}
*123123 - price example

For Example:
//for stop-price=500
{"v":1,"c":{"t":"sp","v":{"p":500}}}

More detailed:

order example with attachment:
https://pastebin.com/upv6qUaq
stop-price: 9747200
json: {"v":1,"c":{"t":"sp","v":{"p":9747200}}}
b58 repr: 77rnoyFX5BDr15hqZiUtgXKSN46zsbHHQjVNrTMLZcjtEfYWLSpnpmJ
order version should be 4 or higher

------------------------------------------------

The mentioned above information is also available in the telegram chat: https://t.me/c/1670109218/94

What is the date for adding limit orders to the CCXT library? 
Is there anything we can do to help?

Please be reminded that we have paid for annual support for the CCXT library (see letter dated July 18, 2022) without any delay. We expect the service to be delivered accordingly.

We look forward to receiving your reply and looking forward to a productive collaboration

**Comment by NikBuka** [comment]
@kroitor @carlosmiei

**Comment by carlosmiei** [comment]
Hello @NikBuka, we will try to update it asap

**Comment by sc0Vu** [comment]
Hi @NikBuka

I'll submit a PR for this asap.

**Comment by NikBuka** [comment]
@carlosmiei @sc0Vu @pcriadoperez @kroitor Guys, we already waiting for a month for any progress from you (since 6 February). 
Tell me,  when you will support our condition orders?

**Comment by kroitor** [comment]
@NikBuka we're working on it, however, we cannot give an estimate on when it will be finished, it's always "as soon as we possibly can". You can help speed it up by submitting a PR to this method: https://github.com/ccxt/ccxt/blob/master/js/wavesexchange.js#L1261

**Comment by NikBuka** [comment]
@kroitor Could you clarify what exactly should be in PR?

**Comment by kroitor** [comment]
@NikBuka in short, we need to add the logic for conditional orders there and make sure that the signing is working correctly. The logic for conditionals can be copied from the other exchanges like in [binance.js](https://github.com/ccxt/ccxt/blob/master/js/binance.js#L3906) or [okx.js](https://github.com/ccxt/ccxt/blob/master/js/okx.js#L2134). That is what we're working on right now.

**Comment by NikBuka** [comment]
@kroitor @sc0Vu ok, we are waiting for your implementation.

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei Hey, tell me please, what progress is for our task?

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hey, tell me please, what progress on our task?

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hey, tell me please, what progress on our task?

**Comment by carlosmiei** [comment]
> @kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hey, tell me please, what progress on our task?

afaik @sc0Vu is working on it

**Comment by NikBuka** [comment]
@sc0Vu Hi, It's been two weeks, do you have any updates on our task?

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hi, tell me please, what progress is on our task?

**Comment by NikBuka** [comment]
@sc0Vu Hi, It's been two weeks, do you have any updates on our task?

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hello all! **It's been 2 months** since our request, tell me please what has been done within it?

**Comment by sc0Vu** [comment]
Hi @NikBuka 

I'm working on this issue, will open PR ASAP.

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hello all! It's been 2 months since our request, tell me please what has been done within it?

**Comment by carlosmiei** [comment]
@sc0Vu can you update us here pls

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hello all! It's been almost **3 months** since our request, tell me please what has been done within it?

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hello all! It's been **more than 3 months** since our request, tell me please what has been done within it?

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hello all! It's been more than 3 months since our request, tell me please what has been done within it?

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hello all! It's been more than 3 months since our request, tell me please what has been done within it?

**Comment by NikBuka** [comment]
@kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez Hello all! It's been more than 3 months since our request, tell me please what has been done within it?

**Comment by kroitor** [comment]
@NikBuka we're having a technical issue adding the `protobuf` as a dependency, cause CCXT is a zero-dependency library, and the Waves API is the only exchange requiring so many external packages. All other exchanges use conventional JSON/HTTP APIs, which is why Waves.Exchange is an outlier in this case. I mean, why would you want to make the API harder to support, instead of making it easier to support? Can you guys do something about it to simplify it and introduce a regular JSON/HTTP API for us to integrate? It wouldn't take more than a few days if the API was easier to maintain. Otherwise it's taking too long.

**Comment by NikBuka** [comment]
> @NikBuka we're having a technical issue adding the `protobuf` as a dependency, cause CCXT is a zero-dependency library, and the Waves API is the only exchange requiring so many external packages. All other exchanges use conventional JSON/HTTP APIs, which is why Waves.Exchange is an outlier in this case. I mean, why would you want to make the API harder to support, instead of making it easier to support for your own users? Can you guys do something about it to simplify it and introduce a regular JSON/HTTP API for us to integrate? It wouldn't take more than a few days if the API was easier to maintain or if it was in line with the other exchanges. Otherwise it's taking too long, cause it's very unique and requires a lot of external first-order dependencies and dependencies of dependencies as well.

Hi @kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez. Oh, if you had told us about this problem a little earlier (like 4 months ago) and asked us for such edits, than we would have made them as soon as possible.
Now we will make a small edit - will add an HTTP route to get an array of bytes from the order, which will need to be signed later.
We need 1 week.

**Comment by NikBuka** [comment]
Hi @kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez.
We have implemented a method to serialize orders in tesnet, please check it out.
You need to send the order here and you will get dase58 in response.
https://matcher-testnet.waves.exchange/api-docs/index.html#/serialize/serializeOrder

**Comment by kroitor** [comment]
@NikBuka hello! Thank you for this update! We will explore it and will be back to you asap!

**Comment by NikBuka** [comment]
> @NikBuka hello! Thank you for this update! We will explore it and will be back to you asap!

Hello @kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez, do you have any news about our updat?

**Comment by kroitor** [comment]
@NikBuka hey! We're in the process of implementing and testing it according to the spec you posted above, will be back to you with our results soon. Thx!

**Comment by NikBuka** [comment]
Hello @kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez, do you have any news about our update?
It's been almost 3 week.

**Comment by NikBuka** [comment]
Hello @kroitor @sc0Vu @carlosmiei @frosty00 @pcriadoperez, do you have any news about our update?
It's been almost **6 Months**. since our first request

**Comment by sc0Vu** [comment]
Hi @NikBuka 

Quick updates: I can submit stop order with attachment, but the order always got cancelled, would you like to help me?

Here is my order:
```
// request
{"senderPublicKey":"8aUTNqHGCBiubySBRhcS1N6NC5jLczhVcndRfMAuwtkY","matcherPublicKey":"8QUAqtTckM5B8gvcuP7mMswat9SjKUuafJMusEoSn1Gy","assetPair":{"amountAsset":"EGo1BQqGxHsc4nFCi5uzpd8eSbH22xGR8hiw9P44Uviw","priceAsset":""},"orderType":"buy","price":1000000,"amount":1000000,"timestamp":1690177586244,"expiration":1692596786244,"matcherFee":1000000,"priceMode":"fixedDecimals","version":4,"attachment":"77rnoyFX5BDr15hqZiUtgXKSN46zsbHHQjVNrTMLZcLz62mmFKr39FJ","signature":"signaturexxx"} 

// response
{"success":true,"message":{"version":4,"id":"AStDxT8NoTPzTiLixJxjpTaFufeF6KhuDQjPhTWLWXiH","sender":"3MpEdBXtsRHRj2TvZURSb8uLDxzneVbYczW","senderPublicKey":"8aUTNqHGCBiubySBRhcS1N6NC5jLczhVcndRfMAuwtkY","matcherPublicKey":"8QUAqtTckM5B8gvcuP7mMswat9SjKUuafJMusEoSn1Gy","assetPair":{"amountAsset":"EGo1BQqGxHsc4nFCi5uzpd8eSbH22xGR8hiw9P44Uviw","priceAsset":"WAVES"},"orderType":"buy","amount":1000000,"price":1000000,"timestamp":1690177586244,"expiration":1692596786244,"matcherFee":1000000,"signature":"signaturexxx","proofs":["4FwFPSDwpRT5RTHRMvjNbzx33KNDWMW6ka16ryuiSLPzam5FLN1rmXy8bmP44gKxuapAkDURBaPtFGFeHEx6LfkA"],"matcherFeeAssetId":null,"eip712Signature":null,"priceMode":"assetDecimals","attachment":"77rnoyFX5BDr15hqZiUtgXKSN46zsbHHQjVNrTMLZcLz62mmFKr39FJ"},"status":"OrderAccepted"} 

```

BTW, when call cancel order with the cancelled order id, I got `success: true` (BatchCancelCompleted). Is this supposed to be false instead?

```
{"message":[[{"error":9437194,"message":"The order AStDxT8NoTPzTiLixJxjpTaFufeF6KhuDQjPhTWLWXiH is canceled","template":"The order {{id}} is canceled","params":{"id":"AStDxT8NoTPzTiLixJxjpTaFufeF6KhuDQjPhTWLWXiH"},"status":"OrderCancelRejected","success":false}]],"success":true,"status":"BatchCancelCompleted"} 
```

**Comment by NikBuka** [comment]
Hi   @sc0Vu by analogy with example below, you need to make your order, sign and send.
Also you can use https://testnet.wx.network/  or https://testnet.waves.exchange/  for test.

curl -XPOST -s -H 'content-type: application/json;charset=UTF-8' 'https://testnet.waves.exchange/api/v1/forward/matcher/matcher/orderbook' \
  --data-raw '{"orderType":"buy","version":4,"assetPair":{"amountAsset":"EMAMLxDnv3xiz8RXg8Btj33jcEw3wLczL3JKYYmuubpc","priceAsset":"WAVES"},"price":17284560,"amount":100000000,"timestamp":1690447164575,"expiration":1692952764575,"matcherFee":83327580,"matcherPublicKey":"8QUAqtTckM5B8gvcuP7mMswat9SjKUuafJMusEoSn1Gy","senderPublicKey":"8QUAqtTckM5B8gvcuP7mMswat9SjKUuafJMusEoSn1Gy","proofs":["maJqdPEFvJ7wd15U7uwmwpdGA3US2iwYmpyDfEtLhwpA31js8XZ89P5gJp67m3FZwf1Z2kLQX4fc3CqNZBiEazW"],"matcherFeeAssetId":"EMAMLxDnv3xiz8RXg8Btj33jcEw3wLczL3JKYYmuubpc","id":"BR3ZwSL2Pg3Mgofx6xRwfLByArXGQPi932vw5tL4RAPW","priceMode":"assetDecimals","chainId":84,"attachment":"U1GmCKH5jS7hHMmFMQ677SpNwvhUSJJsRkYFTHJmLw1CyCpaJE9UJMRe"}'

**Comment by NikBuka** [comment]
@kroitor @sc0Vu hey guys, when do you plan release with our condition orders?

**Comment by kroitor** [comment]
@NikBuka pardon for the long wait, we've merged wavesexchange conditional orders and released them in CCXT 4.0.64+ 

Please, let us know if you have any difficulties with it.
